src/librustc_codegen_llvm/mir/block.rs

```diff
@@ -588,14 +588,40 @@ impl FunctionCx<'a, 'll, 'tcx> {
                     (&args[..], None)
                 };
 
-                for (i, arg) in first_args.iter().enumerate() {
+                'make_args: for (i, arg) in first_args.iter().enumerate() {
                     let mut op = self.codegen_operand(&bx, arg);
+
                     if let (0, Some(ty::InstanceDef::Virtual(_, idx))) = (i, def) {
-                        if let Pair(data_ptr, meta) = op.val {
-                            llfn = Some(meth::VirtualIndex::from_index(idx)
-                                .get_fn(&bx, meta, &fn_ty));
-                            llargs.push(data_ptr);
-                            continue;
+                        // descend through newtype wrappers until `op` is a builtin pointer to
+                        // `dyn Trait`, e.g. `*const dyn Trait`, `&mut dyn Trait`
+                        'descend_newtypes: while !op.layout.ty.is_unsafe_ptr()
```

Comment:
Now that I'm rebasing on #54183, which has conflicts here, I'm starting to realize that this code isn't very easy to understand. So I'm going to try to do the refactor you suggested and add some comments to explain what's going on, since it's *very* hacky. Essentially we're taking a newtyped pointer, removing all the newtype wrappers to get a `*mut dyn Trait`, `&dyn Trait` or similar, taking away the vtable so it's just a thin pointer, but telling the compiler its type is still `*mut dyn Trait` because it understands that as a special case somewhere else according to @eddyb